### PR TITLE
3398 - Attachments - Reclassify File Tag

### PR DESCRIPTION
**Pull request by milosdes**
https://app.zenhub.com/workspaces/foi-flow-607f0ab6ae54b700111dc733/issues/gh/bcgov/foi-flow/3398

### CODE REVIEW

**Comment by milosdes** [comment]
There are some things here I'd like to review with another dev before merging

File: forms-flow-web/src/components/FOI/customComponents/Attachments/AttachmentModal.js [hunk 1]
```diff
@@ -315,21 +350,32 @@ export default function AttachmentModal({
...
-              </button>
-              :
+              </button> : ''
+            }
+             {
+              modalFor !== 'rename' && modalFor !== 'reclassify' ?
               <button className={`btn-bottom btn-save ${ isSaveDisabled() ? classes.btndisabled : classes.btnenabled }`} disabled={isSaveDisabled()} onClick={handleSave}>
                 {uploadFor === "email" ? "Save Changes" : "Continue"}
-              </button>
+              </button> : ''
```
Comment (milosdes) [inline forms-flow-web/src/components/FOI/customComponents/Attachments/AttachmentModal.js:378]: I refactored the ternary operator here to have more modular access to what is displayed based off of the modalFor value - please check to make sure this doesn't break anything though!

File: forms-flow-web/src/components/FOI/customComponents/Attachments/index.js [hunk 2]
```diff
@@ -485,6 +506,8 @@ const Attachment = React.memo(({indexValue, attachment, handlePopupButtonClick,
     }
   }, [attachment])
 
+  const reclassifyIsDisabled = (attachment.category.includes('>') || attachment.category.toLowerCase().includes('applicant')) ? true : false;
+
```
Comment (milosdes) [inline forms-flow-web/src/components/FOI/customComponents/Attachments/index.js:510]: Spoke with Kyle about this. The .includes('>') is the simplest way to check if a file was uploaded by the system. The 'applicant' type is the only other file type without '>' that is tagged by the system. We decided to simply disable reclassifying files tagged with 'applicant'.
The only downside to this is that files manually tagged as 'applicant' cannot be reclassified, they will need to be reuploaded.

File: request-management-api/request_api/resources/foidocument.py [hunk 3]
```diff
@@ -103,6 +104,45 @@ def post(requestid, documentid, requesttype):
...
+                 result = documentservice().createrequestdocumentversion(requestid, documentid, documentschema, AuthHelper.getuserid(), requesttype)
+                 return {'status': result.success, 'message':result.message,'id':result.identifier} , 200
+            return {'status': False, 'message': 'Something went wrong' }, 500
+        except ValidationError as err:
+                    return {'status': False, 'message':err.messages}, 400
+        except KeyError as err:
+            return {'status': False, 'message': 'KeyError'}, 400
+        except BusinessException as exception:
+            return {'status': exception.status_code, 'message':exception.message}, 500
+
```
Comment (milosdes) [inline request-management-api/request_api/resources/foidocument.py:145]: Please review best way to handle errors for this endpoint. 
When the KeyError exception message was set to to err.messages, it was giving a "KeyError does not have messages attribute" error.

File: request-management-api/request_api/schemas/foidocument.py [hunk 4]
```diff
@@ -16,6 +16,14 @@ class Meta:  # pylint: disable=too-few-public-methods
         unknown = EXCLUDE    
     filename = fields.Str(data_key="filename",required=True,allow_none=False)
 
+class ReclassifyDocumentSchema(Schema):
+    class Meta:  # pylint: disable=too-few-public-methods
+        """Exclude unknown fields in the deserialized output."""
+
+        unknown = EXCLUDE
+    category = fields.Str(data_key="category",required=True,allow_none=False)
+    documentpath = fields.Str(data_key="documentpath",required=True,allow_none=False, validate=[validate.Length(max=1000, error=MAX_EXCEPTION_MESSAGE)])
```
Comment (milosdes) [inline request-management-api/request_api/schemas/foidocument.py:25]: I included the documentpath in the schema for reclassifying as it's needed to complete the task, but this wasn't completely necessary as the value isn't being received by the endpoint.

File: request-management-api/request_api/services/documentservice.py [hunk 5]
```diff
@@ -51,6 +51,23 @@ def createrequestdocumentversion(self, requestid, documentid, documentschema, us
...
+        newlocation = location
+        newlocation[3] = newcategory
+        filename = "/".join(newlocation[1:])
+        newdocumentpath = "/".join(baseurl) + '/' + bucket + '/' + filename
+        moveresponse = storageservice().copy_file(source, bucket, filename)
+        if moveresponse['ResponseMetadata']['HTTPStatusCode'] == 200:
+            return {"status": "success", 'documentpath': newdocumentpath}
+        else:
+            return {"status": "error"}
+
```
Comment (milosdes) [inline request-management-api/request_api/services/documentservice.py:70]: The documentpath here is being parsed based on the assumption that the whole path include 'http://' is passed. 
Also please advise on best way to handle error.

File: forms-flow-web/src/components/FOI/customComponents/Attachments/index.js (hunk 2 above)
Comment (milosdes) [reply forms-flow-web/src/components/FOI/customComponents/Attachments/index.js:510]: Spoke with Nick - use the AttachmentCategory enum here to compare to the attachment category value.

File: forms-flow-web/src/components/FOI/customComponents/Attachments/AttachmentModal.js (hunk 1 above)
Comment (milosdes) [reply forms-flow-web/src/components/FOI/customComponents/Attachments/AttachmentModal.js:378]: Change ternaries to conditional rendering

File: forms-flow-web/src/components/FOI/customComponents/Attachments/index.js [hunk 6]
```diff
@@ -485,6 +506,19 @@ const Attachment = React.memo(({indexValue, attachment, handlePopupButtonClick,
...
+    AttachmentCategories.categorys.forEach((category) => {
+      if (category.name.toLowerCase() === attachment.category.toLowerCase()) {
+        if (category.display.toLowerCase().includes('>') || category.display.toLowerCase().includes('applicant')) {
+          result = true
+        }
+      }
+    })
+    return result
+  }
+  const reclassifyIsDisabled = disableAttachmentsTaggedBySystem(attachment)
```
Comment (milosdes) [inline forms-flow-web/src/components/FOI/customComponents/Attachments/index.js:520]: Changed this as suggested

File: request-management-api/request_api/services/documentservice.py [hunk 7]
```diff
@@ -51,6 +52,23 @@ def createrequestdocumentversion(self, requestid, documentid, documentschema, us
         else:
             return self.createrawdocumentversion(requestid, documentid, documentschema, userid)
 
+    def copyrequestdocumenttonewlocation(self, newcategory, documentpath): #documentpath is full url including https://
+        # for old documentpath
+        baseurl = 'https://' + os.getenv("OSS_S3_HOST")
```
Comment (milosdes) [inline request-management-api/request_api/services/documentservice.py:57]: Changed this

File: request-management-api/request_api/services/documentservice.py [hunk 8]
```diff
@@ -51,6 +52,23 @@ def createrequestdocumentversion(self, requestid, documentid, documentschema, us
...
+        newlocation = location
+        newlocation[3] = newcategory
+        filename = "/".join(newlocation[1:])
+        newdocumentpath = "/".join(baseurl) + '/' + bucket + '/' + filename
+        moveresponse = storageservice().copy_file(source, bucket, filename)
+        if moveresponse['ResponseMetadata']['HTTPStatusCode'] == 200:
+            return {"status": "success", 'documentpath': newdocumentpath}
+        else:
+            return {"status": "error"}
+
```
Comment (milosdes) [inline request-management-api/request_api/services/documentservice.py:71]: It looks like we can't use raise_for_status() here, it's part of the Python requests module, that's not being used to make the S3 request, we're using the boto3 client instead.

Comment (milosdes) [reply request-management-api/request_api/services/documentservice.py:71]: for ref:
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/copy_object.html

File: request-management-api/request_api/services/documentservice.py [hunk 9]
```diff
@@ -51,6 +53,27 @@ def createrequestdocumentversion(self, requestid, documentid, documentschema, us
...
+        newdocumentpath = baseurl + '/' + bucket + '/' + filename
+        try:
+            moveresponse = storageservice().copy_file(source, bucket, filename)
+            if moveresponse['ResponseMetadata']['HTTPStatusCode'] == 200:
+                return {"status": "success", 'documentpath': newdocumentpath}
+            else:
+                raise Exception(moveresponse)
+        except Exception as ex:
+            logging.exception(ex)
+            return {"status": "error"}
```
Comment (milosdes) [inline request-management-api/request_api/services/documentservice.py:75]: Throw an exception if the boto3 client doesn't return a 200 response for copying the file on s3

Comment (nkan-aot) [reply request-management-api/request_api/services/documentservice.py:75]: can use raise for status here, will take fewer lines

**Review by nkan-aot — APPROVED** [review]

**Pull request merged**